File: Leetcode_POTD_2025/7_JULY/1717_Maximum_Score_From_Removing_Substrings.java
class Solution {
    public int maximumGain(String s, int x, int y) {
        int totalPoints = 0;
        StringBuilder tempSbA = new StringBuilder();
        StringBuilder tempSbB = new StringBuilder();
        char firstChar, secondChar;
        int firstGain, secondGain;

        if (x >= y) {
            firstChar = 'a';
            secondChar = 'b';
            firstGain = x;
            secondGain = y;
        } else {
            firstChar = 'b';
            secondChar = 'a';
            firstGain = y;
            secondGain = x;
        }

        for (char currentChar : s.toCharArray()) {
            if (tempSbA.length() > 0 && tempSbA.charAt(tempSbA.length() - 1) == firstChar && currentChar == secondChar) {
                totalPoints += firstGain;
                tempSbA.deleteCharAt(tempSbA.length() - 1);
            } else {
                tempSbA.append(currentChar);
            }
        }

        for (char currentChar : tempSbA.toString().toCharArray()) {
            if (tempSbB.length() > 0 && tempSbB.charAt(tempSbB.length() - 1) == secondChar && currentChar == firstChar) {
                totalPoints += secondGain;
                tempSbB.deleteCharAt(tempSbB.length() - 1);
            } else {
                tempSbB.append(currentChar);
            }
        }
        return totalPoints;
    }
}

File: Leetcode_POTD_2025/7_JULY/1717_Maximum_Score_From_Removing_Substrings.py
class Solution(object):
    def maximumGain(self, s, x, y):
        """
        :type s: str
        :type x: int
        :type y: int
        :rtype: int
        """
        total_points = 0
        
        # Determine which pair to prioritize
        if x >= y:
            first_pair = "ab"
            first_gain = x
            second_pair = "ba"
            second_gain = y
        else:
            first_pair = "ba"
            first_gain = y
            second_pair = "ab"
            second_gain = x
            
        # First pass for the higher-scoring pair
        temp_list = []
        for char_s in s:
            if temp_list and temp_list[-1] == first_pair[0] and char_s == first_pair[1]:
                total_points += first_gain
                temp_list.pop()
            else:
                temp_list.append(char_s)
        
        # Second pass for the lower-scoring pair on the remaining string
        remaining_s = "".join(temp_list)
        temp_list = []
        for char_s in remaining_s:
            if temp_list and temp_list[-1] == second_pair[0] and char_s == second_pair[1]:
                total_points += second_gain
                temp_list.pop()
            else:
                temp_list.append(char_s)
                
        return total_points
